Feature/kafka guide app

gokulavasan opened the pull request:
Adding a section to the docs on publishing messages to Kafka topic.

gokulavasan committed:
Switch cdap-kafka-pack version to 0.1.0

gokulavasan committed:
Merge remote-tracking branch 'origin/develop' into feature/kafka-guide-app

gokulavasan committed:
Adding section on publishing messages through Kafka

gokulavasan commented:
@awholegunch @abaranau Please review.

gokulavasan committed:
Pretty printing bash cmds

awholegunch commented on pom.xml:
```diff
@@ -29,7 +29,7 @@
     <app.main.class>co.cask.cdap.guides.kafka.KafkaIngestionApp</app.main.class>
     <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
     <cdap.version>2.5.1</cdap.version>
```
2.5.2

awholegunch commented on README.md:
```diff
@@ -14,16 +15,18 @@ average size of the messages received. You will:
   [Service](http://docs.cdap.io/cdap/current/en/developers-manual/building-blocks/services.html)
-  to retrieve the analysis results via an HTTP RESTful endpoint.
+  to retrieve the analysis results via an HTTP RESTful endpoint; and 
+- Start a Kafka Server v0.8.x and publish messages to the topic, to which CDAP Application has subscribed.
```
and publish messages to the topic which the CDAP Application is subscribed.

awholegunch commented on README.md:
````diff
@@ -120,12 +123,12 @@ public class KafkaIngestionFlow implements Flow {
 ```
 
-The `KafkaSubFlowlet` subclasses from the `Kafka08ConsumerFlowlet` that is
+The `KafkaConsumerFlowlet` subclasses from the `Kafka08ConsumerFlowlet` that is
````
The `KafkaConsumerFlowlet` subclasses from the `Kafka08ConsumerFlowlet` available in the `cdap-kafka-pack` library:

awholegunch commented on README.md:
```diff
@@ -194,11 +197,13 @@ public class KafkaStatsHandler extends AbstractHttpServiceHandler {
+a Kafka topic must be provided as runtime arguments. You can provide these to the `KafkaConsumerFlowlet` as 
+runtime arguments of the `KafkaIngestionFlow`. The keys of these runtime arguments are 
+(see the [Build and Run Application](#build-and-run-application) section for information on how to pass 
+them to the program at start):
```
(See the [Build and Run Application](#build-and-run-application) section for information on how to pass 
the arguments to the program at the start.) The keys of these runtime arguments are:

awholegunch commented on README.md:
```diff
@@ -221,20 +226,46 @@ If you haven't already started a standalone CDAP installation, start it with the
 
 We can then deploy the application to a standalone CDAP installation and
-start its components (note the runtime arguments as described above in [Configuring the KafkaSubFlowlet](#configuring-the-kafkasubflowlet)):
+start its components (note the runtime arguments as described above in [Configuring the KafkaConsumerFlowlet](#configuring-the-kafkaconsumerflowlet)):
```
note the runtime arguments, as described above in...

awholegunch commented on README.md:
```diff
@@ -221,20 +226,46 @@ If you haven't already started a standalone CDAP installation, start it with the
 
     cdap-cli.sh deploy app target/cdap-kafka-ingest-guide-1.0.0.jar
-    curl http://localhost:10000/v2/apps/KafkaIngestionApp/flows/KafkaIngestionFlow/start -d '{"kafka.zookeeper":"127.0.0.1:1234", "kafka.topic":"MyTopic"}'
+    curl http://localhost:10000/v2/apps/KafkaIngestionApp/flows/KafkaIngestionFlow/start -d '{"kafka.zookeeper":"localhost:2181", "kafka.topic":"MyTopic"}'
```
Is 2181 the default port for kafka.zookeeper? Do we need to point that out?

awholegunch commented on README.md:
```diff
@@ -221,20 +226,46 @@ If you haven't already started a standalone CDAP installation, start it with the
     
-Once the Flow is started, Kafka messages are processed as they are published. You can query for
-the average Kafka message size:
+Once the Flow is started, Kafka messages are processed as they are published. Now, lets start Kafka to publish messages to `MyTopic`.
```
Now, let’s start Kafka publishing messages to `MyTopic`.

awholegunch commented on README.md:
```diff
@@ -221,20 +226,46 @@ If you haven't already started a standalone CDAP installation, start it with the
+
+Publish Messages to a Kakfa topic
+---------------------------------
+If you don't have a Kafka v0.8.x running, you can download the binary at [Kafka 0.8.x Download](http://kafka.apache.org/downloads.html).
```
If you don't have Kafka v0.8.x, you can download the binary at [Kafka 0.8.x Download](http://kafka.apache.org/downloads.html).

awholegunch commented on README.md:
```diff
@@ -221,20 +226,46 @@ If you haven't already started a standalone CDAP installation, start it with the
+Publish Messages to a Kakfa topic
+---------------------------------
+If you don't have a Kafka v0.8.x running, you can download the binary at [Kafka 0.8.x Download](http://kafka.apache.org/downloads.html).
+Make sure you download v0.8.x since this guide is designed to work with that version. 
```
Be sure you download v0.8.x, as this guide is designed to work specifically with that version.

awholegunch commented on README.md:
```diff
@@ -221,20 +226,46 @@ If you haven't already started a standalone CDAP installation, start it with the
+If you don't have a Kafka v0.8.x running, you can download the binary at [Kafka 0.8.x Download](http://kafka.apache.org/downloads.html).
+Make sure you download v0.8.x since this guide is designed to work with that version. 
+
+Follow the instructions on [Kafka v0.8.x Quickstart](https://kafka.apache.org/08/quickstart.html), to publish messages to `MyTopic`.
```
...quickstart.html) to publish messages...

awholegunch commented on README.md:
```diff
@@ -221,20 +226,46 @@ If you haven't already started a standalone CDAP installation, start it with the
+Make sure you download v0.8.x since this guide is designed to work with that version. 
+
+Follow the instructions on [Kafka v0.8.x Quickstart](https://kafka.apache.org/08/quickstart.html), to publish messages to `MyTopic`.
+The instructions are repeated below for your convenience (assuming you have download the binary distribution). 
```
Those instructions are repeated below for your convenience and assume you have downloaded the binary distribution.

awholegunch commented on README.md:
````diff
@@ -221,20 +226,46 @@ If you haven't already started a standalone CDAP installation, start it with the
+$ bin/kafka-console-producer.sh --broker-list localhost:9092 --topic MyTopic
+```
+Once the kafka-console-producer.sh script is invoked, you can type messages on the console and every line is published as a message
+to `MyTopic`. Now, go ahead and publish few messages.
````
Go ahead and publish a few messages. "CDAP and Kafka, working together!"...

awholegunch commented on README.md:
```diff
@@ -221,20 +226,46 @@ If you haven't already started a standalone CDAP installation, start it with the
+
+Query Results
+-------------
+You can query for the average Kafka message size:
```
You can query for the average size of the Kafka messages:

awholegunch commented on README.md:
````diff
@@ -221,20 +226,46 @@ If you haven't already started a standalone CDAP installation, start it with the
+```bash
+$ tar xzf kafka-<VERSION>.tgz
+$ cd kafka-<VERSION>
+#Start Zookeeper Server
````
Perhaps a space after the #, to make reading easier?

awholegunch commented on README.md:
````diff
@@ -221,20 +226,46 @@ If you haven't already started a standalone CDAP installation, start it with the
+Follow the instructions on [Kafka v0.8.x Quickstart](https://kafka.apache.org/08/quickstart.html), to publish messages to `MyTopic`.
+The instructions are repeated below for your convenience (assuming you have download the binary distribution). 
+
+```bash
````
Rather than this syntax, for console lines, I'd recommend leading with four spaces. When it gets converted to RST (and I'm sure it will at some point) it will work out better.

gokulavasan replied:
@awholegunch I didn't quite understand how to do this. Please feel free to push a commit to this branch addressing this comment.

awholegunch commented on README.md:
```diff
@@ -60,9 +63,9 @@ contents of these files described below):
     ./src/main/java/co/cask/cdap/guides/kafka/KafkaIngestionApp.java
     ./src/main/java/co/cask/cdap/guides/kafka/KafkaIngestionFlow.java
-    ./src/main/java/co/cask/cdap/guides/kafka/KafkaMsgCounterFlowlet.java
+    ./src/main/java/co/cask/cdap/guides/kafka/KafkaConsumerFlowlet.java
```
This should be after the pom.xml.

awholegunch commented:
Numerous comments. I can review it again.

gokulavasan committed:
Addressing PR comments

gokulavasan commented:
@awholegunch Thanks for the review. I have addressed all your comments except one. Feel free to modify the PR to address it. Please take a look again!

awholegunch commented on README.md:
````diff
@@ -221,20 +225,46 @@ If you haven't already started a standalone CDAP installation, start it with the
+$ bin/kafka-console-producer.sh --broker-list localhost:9092 --topic MyTopic
+```
+Once the kafka-console-producer.sh script is invoked, you can type messages on the console and every line is published as a message
+to `MyTopic`. Go ahead and publish few messages. "CDAP and Kafka, working together!".
````
publish a few

awholegunch replied:
For this code sample, because it is console rather than java, rather than using the triple-backtick syntax, you can simply start each line with four spaces (before the dollar-signs) and it will become a code fragment automagically. This will translate better later on, I suspect. I don't think 'bash' is recognized by Pygments...

awholegunch commented:
I see another formatting issue, near "Build and Run Application":
The KafkaIngestionApp application can be built and packaged using the Apache Maven command:
should be
The `KafkaIngestionApp` application can be built and packaged using the Apache Maven command:
so that it becomes
The `KafkaIngestionApp` application can be built and packaged using the Apache Maven command:

gokulavasan committed:
Addressing PR comments

gokulavasan commented:
@awholegunch Addressed your comments. Take a look. Thanks!

awholegunch commented:
A couple of comments, otherwise LGTM.

abaranau commented on README.md:
```diff
@@ -221,20 +225,46 @@ If you haven't already started a standalone CDAP installation, start it with the
     
-Once the Flow is started, Kafka messages are processed as they are published. You can query for
-the average Kafka message size:
+Once the Flow is started, Kafka messages are processed as they are published. Now, let's start Kafka publishing messages to `MyTopic`.
```
Now let's send some data to Kafka topic.

awholegunch replied:
Now, let's send data to the Kafka topic.

abaranau commented:
minor comment. Also, I don't think you want to change diagram. The change doesn't help to explain app design, right? ;)

abaranau commented:
other than that, LGTM

abaranau commented:
please merge and let's verify it!

awholegunch commented:
Slight change to Sasha's comment. Otherwise LGTM.

gokulavasan committed:
Addressing PR comments

gokulavasan committed:
Removed kafka_console_producer from app-design

gokulavasan committed:
Flow name in diagram is KafkaIngestionFlow

gokulavasan commented:
@awholegunch Addressed the comment.
@abaranau Modified the app-design image a bit (removed kafka_console_producer) but I think the Kafka Server in the picture provides more clarity. Let me know if you definitely want the image to be changed.
Merging now.
